<!doctype html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<style>
h1,
h2,
h3,
h4,
h5,
h6,
p,
blockquote {
    margin: 0;
    padding: 0;
}
body {
    font-family: "Helvetica Neue", Helvetica, "Hiragino Sans GB", Arial, sans-serif;
    font-size: 13px;
    line-height: 18px;
    color: #737373;
    background-color: white;
    margin: 10px 13px 10px 13px;
}
table {
	margin: 10px 0 15px 0;
	border-collapse: collapse;
}
td,th {	
	border: 1px solid #ddd;
	padding: 3px 10px;
}
th {
	padding: 5px 10px;	
}

a {
    color: #0069d6;
}
a:hover {
    color: #0050a3;
    text-decoration: none;
}
a img {
    border: none;
}
p {
    margin-bottom: 9px;
}
h1,
h2,
h3,
h4,
h5,
h6 {
    color: #404040;
    line-height: 36px;
}
h1 {
    margin-bottom: 18px;
    font-size: 30px;
}
h2 {
    font-size: 24px;
}
h3 {
    font-size: 18px;
}
h4 {
    font-size: 16px;
}
h5 {
    font-size: 14px;
}
h6 {
    font-size: 13px;
}
hr {
    margin: 0 0 19px;
    border: 0;
    border-bottom: 1px solid #ccc;
}
blockquote {
    padding: 13px 13px 21px 15px;
    margin-bottom: 18px;
    font-family:georgia,serif;
    font-style: italic;
}
blockquote:before {
    content:"\201C";
    font-size:40px;
    margin-left:-10px;
    font-family:georgia,serif;
    color:#eee;
}
blockquote p {
    font-size: 14px;
    font-weight: 300;
    line-height: 18px;
    margin-bottom: 0;
    font-style: italic;
}
code, pre {
    font-family: Monaco, Andale Mono, Courier New, monospace;
}
code {
    background-color: #fee9cc;
    color: rgba(0, 0, 0, 0.75);
    padding: 1px 3px;
    font-size: 12px;
    -webkit-border-radius: 3px;
    -moz-border-radius: 3px;
    border-radius: 3px;
}
pre {
    display: block;
    padding: 14px;
    margin: 0 0 18px;
    line-height: 16px;
    font-size: 11px;
    border: 1px solid #d9d9d9;
    white-space: pre-wrap;
    word-wrap: break-word;
}
pre code {
    background-color: #fff;
    color:#737373;
    font-size: 11px;
    padding: 0;
}
sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}
* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:10px auto;
    }
}
@media print {
	body,code,pre code,h1,h2,h3,h4,h5,h6 {
		color: black;
	}
	table, pre {
		page-break-inside: avoid;
	}
}
</style>
<title>Prerequisites</title>

</head>
<body>
<h1>bio-SAMtools Basic Tutorial</h1>

<h2>Introduction</h2>

<p>bio-SAMtools is a Ruby binding to the popular <a href="http://SAMtools.sourceforge.net/">SAMtools</a> library, and provides access to individual read alignments as well as BAM files, reference sequence and pileup information.</p>

<h2>Installation</h2>

<p>Installation of bio-SAMtools is very straightforward, and is
accomplished with the Ruby gems command. All you need is an internet
connection.</p>

<h3>Prerequisites</h3>

<p>bio-SAMtools relies on the following other rubygems:</p>

<ul>
<li><a href="http://rubygems.org/gems/bio">bio >= 1.4.2</a></li>
<li><a href="https://rubygems.org/gems/bio-svgenes">bio-svgenes >= 0.4.1</a></li>
</ul>


<p>Once these are installed, bio-SAMtools can be installed with</p>

<pre><code>sudo gem install bio-SAMtools
</code></pre>

<p>It should then be easy to test whether installation went well. Start
interactive Ruby (IRB) in the terminal, and type
<code>require 'bio-SAMtools'</code> if the terminal returns <code>true</code> then all is
well.</p>

<pre><code>$ irb
&gt;&gt; require 'bio-SAMtools'
=&gt; true
</code></pre>

<h2>Working with BAM files</h2>

<h3>Creating a new SAM object</h3>

<p>A SAM object represents the alignments in the BAM file. BAM files (and hence SAM objects here) are what most of SAMtools methods operate on and are very straightforward to create. You will need a sorted BAM file, to access the alignments and a reference sequence in FASTA format to use the reference sequence. The object can be created and opened as follows:</p>

<pre><code>bam = Bio::DB::Sam.new(:bam=&gt;"my_sorted.bam", :fasta=&gt;'ref.fasta')
</code></pre>

<p>Opening the file needs only to be done once for multiple operations on
it, access to the alignments is random so you don't need to loop over
the entries in the file.</p>

<h3>Getting Reference Sequence</h3>

<p>The reference is accessed using reference
name, start, end in 1-based co-ordinates. A standard Ruby String object is returned.</p>

<pre><code>sequence_fragment = bam.fetch_reference("Chr1", 1, 100)
</code></pre>

<p>The output from this would be the raw sequence as a string, e.g.</p>

<pre><code>cctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaacccta
</code></pre>

<p>A reference sequence can be returned as a Bio::Sequence::NA object buy the use of :as_bio => true</p>

<pre><code>sequence_fragment = bam.fetch_reference("Chr1", 1, 100, :as_bio =&gt; true)
</code></pre>

<p>The output from this would be a Bio::Sequence::NA object, which provides a fasta-formatted string when printed</p>

<pre><code>&gt;Chr1:1-100
cctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaacccta
</code></pre>

<h3>Alignment Objects</h3>

<p>The individual alignments represent a single read and are returned as
Bio::DB::Alignment objects. These have numerous methods of their own,
using <code>require 'pp'</code> will allow you to check the attributes contained in
each object. Here is an example alignment object. Remember <code>@</code>
represents a Ruby instance variable and can be accessed as any other
method. Thus the <code>@is_mapped</code> attribute of an object <code>a</code> is accessed
<code>a.is_mapped</code></p>

<pre><code>require 'pp'
pp an_alignment_object ##some Bio::DB::Alignment object
#&lt;Bio::DB::Alignment:0x101113f80
@al=#&lt;Bio::DB::SAM::Tools::Bam1T:0x101116a50&gt;,
@calend=4067,
@cigar="76M",
@failed_quality=false,
@first_in_pair=false,
@flag=163,
@is_duplicate=false,
@is_mapped=true,
@is_paired=true,
@isize=180,
@mapq=60,
@mate_strand=false,
@mate_unmapped=false,
@mpos=4096,
@mrnm="=",
@pos=3992,
@primary=true,
@qlen=76,
@qname="HWI-EAS396_0001:7:115:17904:15958#0",
@qual="IIIIIIIIIIIIHHIHGIHIDGGGG...",
@query_strand=true,
@query_unmapped=false,
@rname="1",
@second_in_pair=true,
@seq="ACAGTCCAGTCAAAGTACAAATCGAG...",
@tags=
    {"MD"=&gt;#&lt;Bio::DB::Tag:0x101114ed0 @tag="MD", @type="Z", @value="76"&gt;,
     "XO"=&gt;#&lt;Bio::DB::Tag:0x1011155d8 @tag="XO", @type="i", @value="0"&gt;,
     "AM"=&gt;#&lt;Bio::DB::Tag:0x101116280 @tag="AM", @type="i", @value="37"&gt;,
     "X0"=&gt;#&lt;Bio::DB::Tag:0x101115fb0 @tag="X0", @type="i", @value="1"&gt;,
     "X1"=&gt;#&lt;Bio::DB::Tag:0x101115c68 @tag="X1", @type="i", @value="0"&gt;,
     "XG"=&gt;#&lt;Bio::DB::Tag:0x101115240 @tag="XG", @type="i", @value="0"&gt;,
     "SM"=&gt;#&lt;Bio::DB::Tag:0x1011162f8 @tag="SM", @type="i", @value="37"&gt;,
     "XT"=&gt;#&lt;Bio::DB::Tag:0x1011162a8 @tag="XT", @type="A", @value="U"&gt;,
     "NM"=&gt;#&lt;Bio::DB::Tag:0x101116348 @tag="NM", @type="i", @value="0"&gt;,
     "XM"=&gt;#&lt;Bio::DB::Tag:0x101115948 @tag="XM", @type="i", @value="0"&gt;}&gt;
</code></pre>

<h3>Getting Alignments</h3>

<p>Alignments can be obtained one at a time by looping over a specified region using the <code>fetch()</code> function.</p>

<pre><code>bam.fetch("Chr1",3000,4000).each do |alignment|
    #do something with the alignment...
end
</code></pre>

<p>A separate method <code>fetch_with_function()</code> allows you to pass a block (or
a Proc object) to the function for efficient calculation. This example takes
an alignment object and returns an array of sequences which exactly match the reference.</p>

<pre><code>#an array to hold the matching sequences
exact_matches = []

matches = Proc.new do |a|
    #get the length of each read
    len = a.seq.length
    #get the cigar string
    cigar = a.cigar
    #create a cigar string which represents a full-length match
    cstr = len.to_s &lt;&lt; "M"
    if cigar == cstr
        #add the current sequence to the array if it qualifies
        exact_matches &lt;&lt; a.seq
    end
end

bam.fetch_with_function("Chr1", 100, 500, &amp;matches) 

puts exact_matches
</code></pre>

<h3>Alignment stats</h3>

<p>The SAMtools flagstat method is implemented in bio-SAMtools to quickly examine the number of reads mapped to the reference. This includes the number of paired and singleton reads mapped and also the number of paired-reads that map to different chromosomes/contigs.</p>

<pre><code>bam.flag_stats()
</code></pre>

<p>An example output would be</p>

<pre><code>34672 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 duplicates
33196 + 0 mapped (95.74%:nan%)
34672 + 0 paired in sequencing
17335 + 0 read1
17337 + 0 read2
31392 + 0 properly paired (90.54%:nan%)
31728 + 0 with itself and mate mapped
1468 + 0 singletons (4.23%:nan%)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre>

<h2>Getting Coverage Information</h2>

<h3>Per Base Coverage</h3>

<p>It is easy to get the total depth of reads at a given position, the
<code>chromosome_coverage</code> function is used. This differs from the previous
functions in that a start position and length (rather than end position)
are passed to the function. An array of coverages is returned, the first
position in the array gives the depth of coverage at the given start
position in the genome, the last position in the array gives the depth
of coverage at the given start position plus the length given</p>

<pre><code>coverages = bam.chromosome_coverage("Chr1", 3000, 1000)  #=&gt; [16,16,25,25...]
</code></pre>

<h3>Average Coverage In A Region</h3>

<p>Similarly, average (arithmetic mean) of coverage can be retrieved with the <code>average_coverage</code> method.</p>

<pre><code>coverages = bam.average_coverage("Chr1", 3000, 1000)  #=&gt; 20.287
</code></pre>

<h2>Getting Pileup Information</h2>

<p>Pileup format represents the coverage of reads over a single base in the
reference. Getting a Pileup over a region is very easy. Note that this
is done with <code>mpileup</code> and NOT the now deprecated SAMtools <code>pileup</code>
function. Calling the <code>mpileup</code> method creates an iterator that yields a
Pileup object for each base.</p>

<pre><code>bam.mpileup do |pileup|
    puts pileup.consensus #gives the consensus base from the reads for that postion
end 
</code></pre>

<h3>Caching pileups</h3>

<p>A pileup can be cached, so if you want to execute several operations on the same set of regions, mpilup won't be executed several times. Whenever you finish using a region, call mpileup_clear_cache to free the cache. The argument 'Region' is required, as it will be the key for the underlying hash. We asume that the options (other than the region) are constant. If they are not, the cache mechanism may not be consistent.</p>

<pre><code>#create an mpileup
reg = Bio::DB::Fasta::Region.new
reg.entry = "Chr1"
reg.start = 1
reg.end = 334

bam.mpileup_cached(:r=&gt;reg,:g =&gt; false, :min_cov =&gt; 1, :min_per =&gt;0.2) do |pileup|
    puts pileup.consensus
end
bam.mpileup_clear_cache(reg)
</code></pre>

<h4>Pileup options</h4>

<p>The <code>mpileup</code> function takes a range of parameters to allow SAMtools
level filtering of reads and alignments. They are specified as key =>
value pairs eg</p>

<pre><code>bam.mpileup(:r =&gt; "Chr1:1000-2000", :Q =&gt; 50) do |pileup|
    ##only pileups on Chr1 between positions 1000-2000 are considered, 
    ##bases with Quality Score &lt; 50 are excluded
    ...
end 
</code></pre>

<p>Not all the options SAMtools allows you to pass to mpileup will return a
Pileup object, those that cause mpileup to return BCF/VCF will be
ignored. Specifically these are g,u,e,h,I,L,o,p. The table below lists
the SAMtools flags supported and the symbols you can use to call them in
the mpileup command.</p>

<table><tr><th>SAMtools options</th><th>description</th><th>short symbol</th><th>long symbol</th><th>default</th><th>example</th></tr>
<tr><td>r</td><td>limit retrieval to a region</td><td>:r</td><td>:region</td><td>all positions</td><td>:r => "Chr1:1000-2000"</td></tr>
<tr><td>6</td><td>assume Illumina scaled quality scores</td><td>:six</td><td>:illumina_quals</td><td>false</td><td>:six => true</td></tr>
<tr><td>A</td><td>count anomalous read pairs scores</td><td>:A</td><td>:count_anomalous</td><td>false</td><td>:A => true</td></tr>
<tr><td>B</td><td>disable BAQ computation</td><td>:B</td><td>:no_baq</td><td>false</td><td>:no_baq => true</td></tr>
<tr><td>C</td><td>parameter for adjusting mapQ</td><td>:C</td><td>:adjust_mapq</td><td>0</td><td>:C => 25</td></tr>
<tr><td>d</td><td>max per-BAM depth to avoid excessive memory usage</td><td>:d</td><td>:max_per_bam_depth</td><td>250</td><td>:d => 123</td></tr>
<tr><td>E</td><td>extended BAQ for higher sensitivity but lower specificity</td><td>:E</td><td>:extended_baq</td><td>false</td><td>:E => true</td></tr>
<tr><td>G</td><td>exclude read groups listed in FILE</td><td>:G</td><td>:exclude_reads_file</td><td>false</td><td>:G => my_file.txt</td></tr>
<tr><td>l</td><td>list of positions (chr pos) or regions (BED)</td><td>:l</td><td>:list_of_positions</td><td>false</td><td>:l => my_posns.bed</td></tr>
<tr><td>M</td><td>cap mapping quality at value</td><td>:M</td><td>:mapping_quality_cap</td><td>60</td><td>:M => 40 </td></tr>
<tr><td>R</td><td>ignore RG tags</td><td>:R</td><td>:ignore_rg</td><td>false</td><td>:R => true </td></tr>
<tr><td>q</td><td>skip alignments with mapping quality smaller than value</td><td>:q</td><td>:min_mapping_quality</td><td>0</td><td>:q => 30 </td></tr>
<tr><td>Q</td><td>skip bases with base quality smaller than value</td><td>:Q</td><td>:imin_base_quality</td><td>13</td><td>:Q => 30</td></tr>
</table>


<h2>Coverage Plots</h2>

<p>You can create images that represent read coverage over binned regions of the reference sequence. The output format is svg. A number of parameters can be changed to alter the style of the plot. In the examples below the bin size and fill_color have been used to create plots with different colours and bar widths.</p>

<p>The following lines of code...</p>

<pre><code>bam.plot_coverage("Chr1", 201, 2000, :bin=&gt;20, :svg =&gt; "out2.svg", :fill_color =&gt; '#F1A1B1')
bam.plot_coverage("Chr1", 201, 2000, :bin=&gt;50, :svg =&gt; "out.svg", :fill_color =&gt; '#99CCFF')
bam.plot_coverage("Chr1", 201, 1000, :bin=&gt;250, :svg =&gt; "out3.svg", :fill_color =&gt; '#33AD5C', :stroke =&gt; '#33AD5C')
</code></pre>

<p><img src="http://ethering.github.io/bio-samtools/images/out2.svg" alt="Coverage plot 1" />
<img src="http://ethering.github.io/bio-samtools/images/out.svg" alt="Coverage plot 2" />
<img src="http://ethering.github.io/bio-samtools/images/out3.svg" alt="Coverage plot 2" /></p>

<h1>VCF methods</h1>

<p>For enhanced snp calling, we've included a VCF class which reflects each non-metadata line of a VCF file.
The VCF class returns the eight fixed fields present in VCF files, namely chromosome, position, ID, reference base, alt bases, alt quality score, filter and info along with the genotype fields, format and samples. This information allows the comparison of variants and their genotypes across any number of samples.
The following code takes a number of VCF objects and examines them for homozygous alt (1/1) SNPs</p>

<pre><code>vcfs = []
vcfs &lt;&lt; vcf1 = Bio::DB::Vcf.new("20 14370   rs6054257   G   A   29  0   NS=3;DP=14;AF=0.5;DB;H2 GT:GQ:DP:HQ 0|0:48:1:51,51  1|0:48:8:51,51  1/1:43:5:-1,-1") #from a 3.3 vcf file
vcfs &lt;&lt; vcf2 = Bio::DB::Vcf.new("19 111 .   A   C   9.6 .   .   GT:HQ   0|0:10,10   0/0:10,10   0/1:3,3") #from a 4.0 vcf file
vcfs &lt;&lt; vcf3 = Bio::DB::Vcf.new("20 14380   rs6054257   G   A   29  PASS    NS=3;DP=14;AF=0.5;DB;H2 GT:GQ:DP:HQ 0|0:48:1:51,51  1|0:48:8:51,51  1/1:43:5:.,") #from a 4.0 vcf file

vcfs.each do |vcf|
    vcf.samples.each do |sample|
        genotype = "#{sample[1]['GT']}"
        if genotype == '1/1' or genotype == '1|1'
            print vcf.chrom, " "
            puts vcf.pos
        end
    end
end

=&gt; 20 14370
=&gt; 20 14380
</code></pre>

<h2>Tests</h2>

<p>The easiest way to run the built-in unit tests is to change to the
bio-SAMtools source directory and running 'rake test'</p>

<p>Each test file tests different aspects of the code.</p>
</body>
</html>