<!DOCTYPE html>

<html>
<head>
<meta charset="UTF-8">

<title>module Bio - Pileup</title>

<link href="./fonts.css" rel="stylesheet">
<link href="./rdoc.css" rel="stylesheet">

<script type="text/javascript">
  var rdoc_rel_prefix = "./";
</script>

<script src="./js/jquery.js"></script>
<script src="./js/navigation.js"></script>
<script src="./js/search_index.js"></script>
<script src="./js/search.js"></script>
<script src="./js/searcher.js"></script>
<script src="./js/darkfish.js"></script>


<body id="top" role="document" class="module">
<nav role="navigation">
  <div id="project-navigation">
    <div id="home-section" role="region" title="Quick navigation" class="nav-section">
  <h2>
    <a href="./index.html" rel="home">Home</a>
  </h2>

  <div id="table-of-contents-navigation">
    <a href="./table_of_contents.html#pages">Pages</a>
    <a href="./table_of_contents.html#classes">Classes</a>
    <a href="./table_of_contents.html#methods">Methods</a>
  </div>
</div>

    <div id="search-section" role="search" class="project-section initially-hidden">
  <form action="#" method="get" accept-charset="utf-8">
    <div id="search-field-wrapper">
      <input id="search-field" role="combobox" aria-label="Search"
             aria-autocomplete="list" aria-controls="search-results"
             type="text" name="search" placeholder="Search" spellcheck="false"
             title="Type to search, Up and Down to navigate, Enter to load">
    </div>

    <ul id="search-results" aria-label="Search Results"
        aria-busy="false" aria-expanded="false"
        aria-atomic="false" class="initially-hidden"></ul>
  </form>
</div>

  </div>

  
<div class="nav-section">
  <h3>Table of Contents</h3>

  <ul class="link-list" role="directory">
    <li><a href="#module-Bio-label-Bio%3A%3ADB%3A%3APileup+">Bio::DB::Pileup </a>
    <li><a href="#module-Bio-label-Vcf+">Vcf </a>
  </ul>
</div>


  <div id="class-metadata">
    
    
    
    
    
  </div>
</nav>

<main role="main" aria-labelledby="module-Bio">
  <h1 id="module-Bio" class="module">
    module Bio
  </h1>

  <section class="description">
    
<h1 id="module-Bio-label-Bio%3A%3ADB%3A%3APileup+"><a href="Bio/DB/Pileup.html">Bio::DB::Pileup</a> <span><a href="#module-Bio-label-Bio%3A%3ADB%3A%3APileup+">&para;</a> <a href="#documentation">&uarr;</a></span></h1>

<p>A class representing information in SAMTools pileup format</p>
<dl class="rdoc-list note-list"><dt>Author
<dd>
<p>Dan MacLean (dan.maclean@tsl.ac.uk)</p>
</dd></dl>

<p>Pileup is described at <a
href="http://sourceforge.net/apps/mediawiki/samtools/index.php?title=SAM_FAQ#I_do_not_understand_the_columns_in_the_pileup_output">sourceforge.net/apps/mediawiki/samtools/index.php?title=SAM_FAQ#I_do_not_understand_the_columns_in_the_pileup_output</a>.
Briefly (when you invoke pileup with the -c option):</p>
<ul><li>
<p>1 reference sequence name</p>
</li><li>
<p>2 reference coordinate</p>
</li><li>
<p>(3) reference base, or `*&#39; for an indel line</p>
</li><li>
<p>(4) genotype where heterozygotes are encoded in the IUB code: M=A/C, R=A/G,
W=A/T, S=C/G, Y=C/T and K=G/T; indels are indicated by, for example,   +A,
-A   or +CC/-C. There is no difference between */+A or +A/*.</p>
</li><li>
<p>(5) Phred-scaled likelihood that the genotype is wrong, which is also
called `consensus quality&#39;.</p>
</li><li>
<p>(6) Phred-scaled likelihood that the genotype is identical to the
reference, which is also called `SNP quality&#39;. Suppose the reference
base is A and in alignment we see 17 G and 3 A. We will get a low consensus
quality because it is difficult to distinguish an A/G heterozygote from a
G/G homozygote. We will get a high SNP quality, though, because the
evidence of a SNP is very strong.</p>
</li><li>
<p>(7) root mean square (RMS) mapping quality</p>
</li><li>
<p>8 # reads covering the position</p>
</li><li>
<p>9 read bases at a SNP line (check the manual page for more information);
the 1st indel allele otherwise</p>
</li><li>
<p>10 base quality at a SNP line; the 2nd indel allele otherwise</p>
</li><li>
<p>(11) indel line only: # reads directly supporting the 1st indel allele</p>
</li><li>
<p>(12) indel line only: # reads directly supporting the 2nd indel allele</p>
</li><li>
<p>(13) indel line only: # reads supporting a third indel allele</p>
</li></ul>

<p>If pileup is invoked without `-c&#39;, indel lines and columns between 3
and 7 inclusive will not be outputted.</p>

<p>NB mpileup uses the 6 column output format eg
“seq2t151tGtGt36t0t99t12t.….……At:9&lt;;;7=&lt;&lt;&lt;&lt;&lt;” Pileup
provides accessors for all columns (6 or 10 column format) and a few other
useful methods</p>

<p>require &#39;rubygems&#39; require&#39;ffi&#39; require
&#39;bio/db/sam/bam&#39;</p>

<h1 id="module-Bio-label-Vcf+">Vcf <span><a href="#module-Bio-label-Vcf+">&para;</a> <a href="#documentation">&uarr;</a></span></h1>

<p>A class representing information in Variant Call Format Forked from vcfruby
at <a
href="https://github.com/jesserod/vcfruby">github.com/jesserod/vcfruby</a> 
Modified and tests written by Dan MacLean (dan.maclean@tsl.ac.uk) VCF is
described at <a
href="http://www.1000genomes.org/node/101">www.1000genomes.org/node/101</a></p>

  </section>

  
  
  
  <section id="5Buntitled-5D" class="documentation-section">
    

    

    

    

    
  </section>
</main>


<footer id="validator-badges" role="contentinfo">
  <p><a href="http://validator.w3.org/check/referer">Validate</a>
  <p>Generated by <a href="http://rdoc.rubyforge.org">RDoc</a> 4.1.1.
  <p>Based on <a href="http://deveiate.org/projects/Darkfish-Rdoc/">Darkfish</a> by <a href="http://deveiate.org">Michael Granger</a>.
</footer>

