<!DOCTYPE html>

<html>
<head>
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">

<title>class Bio::DB::Sam - Pileup</title>

<link type="text/css" media="screen" href="../../rdoc.css" rel="stylesheet">

<script type="text/javascript">
  var rdoc_rel_prefix = "../../";
</script>

<script type="text/javascript" charset="utf-8" src="../../js/jquery.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/navigation.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/search_index.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/search.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/searcher.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/darkfish.js"></script>


<body id="top" class="class">
<nav id="metadata">
  <nav id="home-section" class="section">
  <h3 class="section-header">
    <a href="../../index.html">Home</a>
    <a href="../../table_of_contents.html#classes">Classes</a>
    <a href="../../table_of_contents.html#methods">Methods</a>
  </h3>
</nav>


  <nav id="search-section" class="section project-section" class="initially-hidden">
  <form action="#" method="get" accept-charset="utf-8">
    <h3 class="section-header">
      <input type="text" name="search" placeholder="Search" id="search-field"
             title="Type to search, Up and Down to navigate, Enter to load">
    </h3>
  </form>

  <ul id="search-results" class="initially-hidden"></ul>
</nav>


  

  <div id="file-metadata">
    <nav id="file-list-section" class="section">
  <h3 class="section-header">Defined In</h3>
  <ul>
    <li>lib/bio/db/sam.rb
  </ul>
</nav>

    
  </div>

  <div id="class-metadata">
    
    <nav id="parent-class-section" class="section">
  <h3 class="section-header">Parent</h3>
  
  <p class="link">Object
  
</nav>

    
    
    <!-- Method Quickref -->
<nav id="method-list-section" class="section">
  <h3 class="section-header">Methods</h3>

  <ul class="link-list">
    
    <li ><a href="#method-c-docs">::docs</a>
    
    <li ><a href="#method-c-new">::new</a>
    
    <li ><a href="#method-i-average_coverage">#average_coverage</a>
    
    <li ><a href="#method-i-calmd">#calmd</a>
    
    <li ><a href="#method-i-cat">#cat</a>
    
    <li ><a href="#method-i-chromosome_coverage">#chromosome_coverage</a>
    
    <li ><a href="#method-i-depth">#depth</a>
    
    <li ><a href="#method-i-faidx">#faidx</a>
    
    <li ><a href="#method-i-fetch">#fetch</a>
    
    <li ><a href="#method-i-fetch_reference">#fetch_reference</a>
    
    <li ><a href="#method-i-fetch_with_function">#fetch_with_function</a>
    
    <li ><a href="#method-i-fix_mates">#fix_mates</a>
    
    <li ><a href="#method-i-fixmate">#fixmate</a>
    
    <li ><a href="#method-i-flag_stats">#flag_stats</a>
    
    <li ><a href="#method-i-flagstat">#flagstat</a>
    
    <li ><a href="#method-i-idxstats">#idxstats</a>
    
    <li ><a href="#method-i-index">#index</a>
    
    <li ><a href="#method-i-index_stats">#index_stats</a>
    
    <li ><a href="#method-i-merge">#merge</a>
    
    <li ><a href="#method-i-mpileup">#mpileup</a>
    
    <li ><a href="#method-i-open">#open</a>
    
    <li ><a href="#method-i-phase">#phase</a>
    
    <li ><a href="#method-i-reheader">#reheader</a>
    
    <li ><a href="#method-i-remove_duplicates">#remove_duplicates</a>
    
    <li ><a href="#method-i-rmdup">#rmdup</a>
    
    <li ><a href="#method-i-sort">#sort</a>
    
    <li ><a href="#method-i-targetcut">#targetcut</a>
    
    <li ><a href="#method-i-tview">#tview</a>
    
    <li ><a href="#method-i-view">#view</a>
    
  </ul>
</nav>

  </div>

  <div id="project-metadata">
    <nav id="fileindex-section" class="section project-section">
  <h3 class="section-header">Pages</h3>

  <ul>
  
    <li class="file"><a href="../../LICENSE_txt.html">LICENSE</a>
  
  </ul>
</nav>

    <nav id="classindex-section" class="section project-section">
  <h3 class="section-header">Class and Module Index</h3>

  <ul class="link-list">
  
    <li><a href="../../Bio.html">Bio</a>
  
    <li><a href="../../Bio/DB.html">Bio::DB</a>
  
    <li><a href="../../Bio/DB/Alignment.html">Bio::DB::Alignment</a>
  
    <li><a href="../../Bio/DB/Pileup.html">Bio::DB::Pileup</a>
  
    <li><a href="../../Bio/DB/SAM.html">Bio::DB::SAM</a>
  
    <li><a href="../../Bio/DB/SAM/Library.html">Bio::DB::SAM::Library</a>
  
    <li><a href="../../Bio/DB/SAM/Tools.html">Bio::DB::SAM::Tools</a>
  
    <li><a href="../../Bio/DB/Sam.html">Bio::DB::Sam</a>
  
    <li><a href="../../Bio/DB/Tag.html">Bio::DB::Tag</a>
  
    <li><a href="../../Bio/DB/Vcf.html">Bio::DB::Vcf</a>
  
  </ul>
</nav>

  </div>
</nav>

<div id="documentation">
  <h1 class="class">class Bio::DB::Sam</h1>

  <div id="description" class="description">
    
  </div><!-- description -->

  
  
  
  <section id="5Buntitled-5D" class="documentation-section">
    

    

    

    
    <!-- Attributes -->
    <section id="attribute-method-details" class="method-section section">
      <h3 class="section-header">Attributes</h3>

      
      <div id="attribute-i-bam" class="method-detail">
        <div class="method-heading attribute-method-heading">
          <span class="method-name">bam</span><span
            class="attribute-access-type">[RW]</span>
        </div>

        <div class="method-description">
        
        
        
        </div>
      </div>
      
      <div id="attribute-i-bcftools" class="method-detail">
        <div class="method-heading attribute-method-heading">
          <span class="method-name">bcftools</span><span
            class="attribute-access-type">[RW]</span>
        </div>

        <div class="method-description">
        
        
        
        </div>
      </div>
      
      <div id="attribute-i-fasta" class="method-detail">
        <div class="method-heading attribute-method-heading">
          <span class="method-name">fasta</span><span
            class="attribute-access-type">[RW]</span>
        </div>

        <div class="method-description">
        
        
        
        </div>
      </div>
      
      <div id="attribute-i-last_command" class="method-detail">
        <div class="method-heading attribute-method-heading">
          <span class="method-name">last_command</span><span
            class="attribute-access-type">[RW]</span>
        </div>

        <div class="method-description">
        
        
        
        </div>
      </div>
      
      <div id="attribute-i-samtools" class="method-detail">
        <div class="method-heading attribute-method-heading">
          <span class="method-name">samtools</span><span
            class="attribute-access-type">[RW]</span>
        </div>

        <div class="method-description">
        
        
        
        </div>
      </div>
      
    </section><!-- attribute-method-details -->
    

    <!-- Methods -->
    
     <section id="public-class-5Buntitled-5D-method-details" class="method-section section">
      <h3 class="section-header">Public Class Methods</h3>

    
      <div id="method-c-docs" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">docs</span><span
            class="method-args">(program, command)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>program  - one of &#39;samtools&#39; &#39;bcftools&#39;</p>
</li><li>
<p>command - one of the commands relevant to the program</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="docs-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 349</span>
<span class="ruby-keyword">def</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">docs</span>(<span class="ruby-identifier">program</span>, <span class="ruby-identifier">command</span>)
  <span class="ruby-keyword">return</span> <span class="ruby-string">&quot;program must be &#39;samtools&#39; or &#39;bcftools&#39;&quot;</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">not</span> [<span class="ruby-string">&#39;samtools&#39;</span>, <span class="ruby-string">&#39;bcftools&#39;</span>].<span class="ruby-identifier">include?</span> <span class="ruby-identifier">program</span>
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{program} #{command}&quot;</span>
  <span class="ruby-node">%x#{command}`</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- docs-source -->
          
        </div>

        

        
      </div><!-- docs-method -->

    
      <div id="method-c-new" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">new</span><span
            class="method-args">(args)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Creates a new <a href="Sam.html">Bio::DB::Sam</a> object</p>
<ul><li>
<p>fasta [String] - the path to the Fasta reference sequence</p>
</li><li>
<p>bam [String] - path to bam files</p>
</li><li>
<p>samtools [String] - path to alternative installation of samtools</p>
</li><li>
<p>bcftools [String] - path to alternative installation of bcftools</p>
</li><li>
<p>returns [Bio::DB::Sam] a new `Bio::DB::Sam` object</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="new-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 13</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">initialize</span>(<span class="ruby-identifier">args</span>)
  <span class="ruby-ivar">@fasta</span> = <span class="ruby-identifier">args</span>[<span class="ruby-value">:fasta</span>]
  <span class="ruby-ivar">@bam</span> = <span class="ruby-identifier">args</span>[<span class="ruby-value">:bam</span>]
  <span class="ruby-ivar">@samtools</span> = <span class="ruby-identifier">args</span>[<span class="ruby-value">:samtools</span>] <span class="ruby-operator">||</span> <span class="ruby-constant">File</span>.<span class="ruby-identifier">join</span>(<span class="ruby-constant">File</span>.<span class="ruby-identifier">expand_path</span>(<span class="ruby-constant">File</span>.<span class="ruby-identifier">dirname</span>(<span class="ruby-keyword">__FILE__</span>)),<span class="ruby-string">&#39;sam&#39;</span>,<span class="ruby-string">&#39;external&#39;</span>,<span class="ruby-string">&#39;samtools&#39;</span>)
  <span class="ruby-ivar">@bcftools</span> = <span class="ruby-identifier">args</span>[<span class="ruby-value">:bcftools</span>] <span class="ruby-operator">||</span> <span class="ruby-constant">File</span>.<span class="ruby-identifier">join</span>(<span class="ruby-constant">File</span>.<span class="ruby-identifier">expand_path</span>(<span class="ruby-constant">File</span>.<span class="ruby-identifier">dirname</span>(<span class="ruby-keyword">__FILE__</span>)),<span class="ruby-string">&#39;sam&#39;</span>,<span class="ruby-string">&#39;external&#39;</span>,<span class="ruby-string">&#39;bcftools&#39;</span>)
  
  <span class="ruby-ivar">@files</span> = [<span class="ruby-ivar">@files</span>] <span class="ruby-keyword">if</span> <span class="ruby-ivar">@files</span>.<span class="ruby-identifier">instance_of?</span>(<span class="ruby-constant">String</span>)
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-keyword">nil</span>
  <span class="ruby-identifier">raise</span> <span class="ruby-constant">ArgumentError</span>, <span class="ruby-string">&quot;Need Fasta and at least one BAM or SAM&quot;</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">not</span> <span class="ruby-ivar">@fasta</span> <span class="ruby-keyword">or</span> <span class="ruby-keyword">not</span> <span class="ruby-ivar">@bam</span>
  <span class="ruby-identifier">raise</span> <span class="ruby-constant">IOError</span>, <span class="ruby-string">&quot;File not found&quot;</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">not</span> <span class="ruby-identifier">files_ok?</span>
  <span class="ruby-ivar">@bams</span> = [<span class="ruby-ivar">@bams</span>] <span class="ruby-keyword">if</span> <span class="ruby-ivar">@bams</span>.<span class="ruby-identifier">instance_of?</span> <span class="ruby-constant">String</span>
  
<span class="ruby-keyword">end</span></pre>
          </div><!-- new-source -->
          
        </div>

        

        
      </div><!-- new-method -->

    
    </section><!-- public-class-method-details -->
  
     <section id="public-instance-5Buntitled-5D-method-details" class="method-section section">
      <h3 class="section-header">Public Instance Methods</h3>

    
      <div id="method-i-average_coverage" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">average_coverage</span><span
            class="method-args">(chr,start,length)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>returns the average coverage over the region queried</p>
<ul><li>
<p>chr - the reference name</p>
</li><li>
<p>start - the start position</p>
</li><li>
<p>length - the length of the region queried</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="average_coverage-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 116</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">average_coverage</span>(<span class="ruby-identifier">chr</span>,<span class="ruby-identifier">start</span>,<span class="ruby-identifier">length</span>)
  <span class="ruby-identifier">arr</span> = <span class="ruby-keyword">self</span>.<span class="ruby-identifier">chromosome_coverage</span>(<span class="ruby-identifier">chr</span>,<span class="ruby-identifier">start</span>,<span class="ruby-identifier">length</span>)
  <span class="ruby-identifier">arr</span>.<span class="ruby-identifier">inject</span>{ <span class="ruby-operator">|</span><span class="ruby-identifier">sum</span>, <span class="ruby-identifier">el</span><span class="ruby-operator">|</span> <span class="ruby-identifier">sum</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">el</span> }.<span class="ruby-identifier">to_f</span> <span class="ruby-operator">/</span> <span class="ruby-identifier">arr</span>.<span class="ruby-identifier">size</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- average_coverage-source -->
          
        </div>

        

        
      </div><!-- average_coverage-method -->

    
      <div id="method-i-calmd" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">calmd</span><span
            class="method-args">(opts={}, &block)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Generate the MD tag. If the MD tag is already present, this command will
give a warning if the MD tag generated is different from the existing tag.
Output <a href="SAM.html">SAM</a> by default.</p>
<ul><li>
<p>A - When used jointly with -r this option overwrites the original base
quality.</p>
</li><li>
<p>e - Convert a the read base to = if it is identical to the aligned
reference base. Indel caller does not support the = bases at the moment.</p>
</li><li>
<p>u - Output uncompressed BAM</p>
</li><li>
<p>b - Output compressed BAM</p>
</li><li>
<p>S - The input is <a href="SAM.html">SAM</a> with header lines</p>
</li><li>
<p>C - [INT] Coefficient to cap mapping quality of poorly mapped reads. See
the pileup command for details. [0]</p>
</li><li>
<p>r - Compute the BQ tag (without -A) or cap base quality by BAQ (with -A).</p>
</li><li>
<p>E - Extended BAQ calculation. This option trades specificity for
sensitivity, though the effect is minor.</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="calmd-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 435</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">calmd</span>(<span class="ruby-identifier">opts</span>={}, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&quot;calmd&quot;</span>,  <span class="ruby-identifier">opts</span>, [<span class="ruby-value">:E</span>, <span class="ruby-value">:e</span>, <span class="ruby-value">:u</span>, <span class="ruby-value">:b</span>, <span class="ruby-value">:S</span>, <span class="ruby-value">:r</span>] )<span class="ruby-operator">+</span> <span class="ruby-string">&quot; &quot;</span> <span class="ruby-operator">+</span> <span class="ruby-ivar">@fasta</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">type</span> = <span class="ruby-value">:text</span> 
  <span class="ruby-identifier">klass</span> = <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">DB</span><span class="ruby-operator">::</span><span class="ruby-constant">Alignment</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-identifier">klass</span>, <span class="ruby-identifier">type</span>, <span class="ruby-keyword">true</span>, <span class="ruby-string">&quot;@&quot;</span>,<span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- calmd-source -->
          
        </div>

        

        
      </div><!-- calmd-method -->

    
      <div id="method-i-cat" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">cat</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>h - header.sam</p>
</li><li>
<p>out -[FILE] out file name</p>
</li><li>
<p>bams -[FILES] or <a href="Sam.html">Bio::DB::Sam</a> list of input bams, or
<a href="Sam.html">Bio::DB::Sam</a> objects</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="cat-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 331</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">cat</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">out</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:out</span>]
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:out</span>)

  <span class="ruby-identifier">bam_list</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:bams</span>].<span class="ruby-identifier">collect</span> <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">b</span><span class="ruby-operator">|</span>
    <span class="ruby-identifier">b</span>.<span class="ruby-identifier">bam</span> <span class="ruby-keyword">rescue</span> <span class="ruby-identifier">b</span>
  <span class="ruby-keyword">end</span>.<span class="ruby-identifier">join</span>(<span class="ruby-string">&#39; &#39;</span>)
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:bams</span>)
  <span class="ruby-identifier">options</span> = <span class="ruby-identifier">commandify</span>(<span class="ruby-identifier">opts</span>, [<span class="ruby-value">:h</span>] )
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} cat #{options} -o #{out} #{bam_list}&quot;</span>
  <span class="ruby-identifier">puts</span> <span class="ruby-identifier">command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)

<span class="ruby-keyword">end</span></pre>
          </div><!-- cat-source -->
          
        </div>

        

        
      </div><!-- cat-method -->

    
      <div id="method-i-chromosome_coverage" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">chromosome_coverage</span><span
            class="method-args">(chr,start,length)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>returns an array of coverage for each location for which there are mapped
reads</p>
<ul><li>
<p>chr - the reference name</p>
</li><li>
<p>start - the start position</p>
</li><li>
<p>length - the length of the region queried</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="chromosome_coverage-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 103</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">chromosome_coverage</span>(<span class="ruby-identifier">chr</span>,<span class="ruby-identifier">start</span>,<span class="ruby-identifier">length</span>)
  <span class="ruby-identifier">result</span> = []
  <span class="ruby-identifier">region</span> = <span class="ruby-node">&quot;#{chr}:#{start}-#{start + length}&quot;</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">mpileup</span>(<span class="ruby-value">:r</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">region</span>) <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">p</span><span class="ruby-operator">|</span>
    <span class="ruby-identifier">result</span> <span class="ruby-operator">&lt;&lt;</span> <span class="ruby-identifier">p</span>.<span class="ruby-identifier">coverage</span>
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">result</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- chromosome_coverage-source -->
          
        </div>

        

        
      </div><!-- chromosome_coverage-method -->

    
      <div id="method-i-depth" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">depth</span><span
            class="method-args">(opts={}) { |split(/\t/)| ... }</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>returns an array for each position with [sequence_name, position, depth]</p>
<ul><li>
<p>b - list of positions or regions in BED format</p>
</li><li>
<p>l - [INT] minQLen</p>
</li><li>
<p>q - [INT] base quality threshold</p>
</li><li>
<p>Q - [INT] mapping quality threshold</p>
</li><li>
<p>r - [chr:from-to] region</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="depth-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 480</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">depth</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&quot;depth&quot;</span>, <span class="ruby-identifier">opts</span>)
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-constant">String</span>) <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">line</span><span class="ruby-operator">|</span>
    <span class="ruby-keyword">yield</span> <span class="ruby-identifier">line</span>.<span class="ruby-identifier">split</span>(<span class="ruby-regexp">/\t/</span>)
  <span class="ruby-keyword">end</span>

<span class="ruby-keyword">end</span></pre>
          </div><!-- depth-source -->
          
        </div>

        

        
      </div><!-- depth-method -->

    
      <div id="method-i-faidx" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">faidx</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Index reference sequence in the FASTA format or extract subsequence from
indexed reference sequence. If no region is specified, faidx will index the
file and create &lt;ref.fasta&gt;.fai on the disk. If regions are
speficified, the subsequences will be retrieved and printed to stdout in
the FASTA format. Options - if a subsequence is required</p>
<ul><li>
<p>chr - [STRING] the reference name of the subsequence</p>
</li><li>
<p>start - [INT] the start position for the subsequence</p>
</li><li>
<p>stop - [INT] the stop position for the subsequence</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="faidx-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 224</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">faidx</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">has_key?</span>(<span class="ruby-value">:chr</span>) <span class="ruby-keyword">and</span> <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">has_key?</span>(<span class="ruby-value">:start</span>) <span class="ruby-keyword">and</span> <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">has_key?</span>(<span class="ruby-value">:stop</span>)
  <span class="ruby-identifier">opts</span>={<span class="ruby-value">:as_bio</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-keyword">false</span>}
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">fetch_reference</span>(<span class="ruby-value">:chr</span>,<span class="ruby-value">:start</span>,<span class="ruby-value">:stop</span>,<span class="ruby-identifier">opts</span>)
  <span class="ruby-keyword">else</span>
    <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} faidx #{@fasta}&quot;</span>
    <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
    <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
  <span class="ruby-keyword">end</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- faidx-source -->
          
        </div>

        

        
      </div><!-- faidx-method -->

    
      <div id="method-i-fetch" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">fetch</span><span
            class="method-args">(chr, start,stop, &block)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>fetches a subsequence and calls code block</p>
<ul><li>
<p>chr - the reference name for the subsequence</p>
</li><li>
<p>start - the start position for the subsequence</p>
</li><li>
<p>stop - the stop position for the subsequence</p>
</li><li>
<p>&amp;block - the the block of code to execute</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="fetch-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 88</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">fetch</span>(<span class="ruby-identifier">chr</span>, <span class="ruby-identifier">start</span>,<span class="ruby-identifier">stop</span>, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
  <span class="ruby-identifier">view</span>(
  <span class="ruby-value">:chr</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">chr</span>,
  <span class="ruby-value">:start</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">start</span>,
  <span class="ruby-value">:stop</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">stop</span>, 
  <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>  
  )
<span class="ruby-keyword">end</span></pre>
          </div><!-- fetch-source -->
          
        </div>

        
        <div class="aliases">
          Also aliased as: <a href="Sam.html#method-i-fetch_with_function">fetch_with_function</a>
        </div>
        

        
      </div><!-- fetch-method -->

    
      <div id="method-i-fetch_reference" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">fetch_reference</span><span
            class="method-args">(chr,start,stop, opts={:as_bio => false})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>fetches a subsequence from a reference genome and option returns it as a
Bio::Sequence::NA object</p>
<ul><li>
<p>chr -  [STRING] the reference name for the subsequence</p>
</li><li>
<p>start - [INT] the start position for the subsequence</p>
</li><li>
<p>stop - [INT] the stop position for the subsequence</p>
</li><li>
<p>as_bio - boolean stating if the returned object should be a
Bio::Sequence::NA object</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="fetch_reference-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 207</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">fetch_reference</span>(<span class="ruby-identifier">chr</span>,<span class="ruby-identifier">start</span>,<span class="ruby-identifier">stop</span>, <span class="ruby-identifier">opts</span>={<span class="ruby-value">:as_bio</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-keyword">false</span>})
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} faidx #{@fasta} #{chr}:#{start}-#{stop}&quot;</span>
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">seq</span> = <span class="ruby-string">&quot;&quot;</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-constant">String</span>, <span class="ruby-value">:text</span> ) {<span class="ruby-operator">|</span><span class="ruby-identifier">line</span><span class="ruby-operator">|</span> <span class="ruby-identifier">seq</span> = <span class="ruby-identifier">seq</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">line</span> <span class="ruby-keyword">unless</span> <span class="ruby-identifier">line</span> <span class="ruby-operator">=~</span> <span class="ruby-regexp">/^&gt;/</span>}
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:as_bio</span>]
    <span class="ruby-identifier">seq</span> = <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">Sequence</span><span class="ruby-operator">::</span><span class="ruby-constant">NA</span>.<span class="ruby-identifier">new</span>(<span class="ruby-identifier">seq</span>).<span class="ruby-identifier">to_fasta</span>(<span class="ruby-node">&quot;#{chr}:#{start}-#{stop}&quot;</span>)
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">seq</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- fetch_reference-source -->
          
        </div>

        

        
      </div><!-- fetch_reference-method -->

    
      <div id="method-i-fetch_with_function" class="method-detail method-alias">
        
        <div class="method-heading">
          <span class="method-name">fetch_with_function</span><span
            class="method-args">(chr, start,stop, &block)</span>
          
        </div>
        

        <div class="method-description">
          
          
          
          

          
        </div>

        

        
        <div class="aliases">
          Alias for: <a href="Sam.html#method-i-fetch">fetch</a>
        </div>
        
      </div><!-- fetch_with_function-method -->

    
      <div id="method-i-fix_mates" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">fix_mates</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Fill in mate coordinates, ISIZE and mate related flags from a name-sorted
alignment</p>
<ul><li>
<p>out_bam name of outfile</p>
</li><li>
<p>r - remove unmapped reads and secondary alignments</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="fix_mates-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 247</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">fix_mates</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-comment">#opts.merge!({:out_index=&gt;nil})</span>
  <span class="ruby-identifier">remove_reads</span> = <span class="ruby-string">&quot;&quot;</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:r</span>]
    <span class="ruby-identifier">remove_reads</span> = <span class="ruby-string">&quot;-r&quot;</span>
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} fixmate #{remove_reads} #{@bam} #{opts[:out_bam]}&quot;</span>
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- fix_mates-source -->
          
        </div>

        
        <div class="aliases">
          Also aliased as: <a href="Sam.html#method-i-fixmate">fixmate</a>
        </div>
        

        
      </div><!-- fix_mates-method -->

    
      <div id="method-i-fixmate" class="method-detail method-alias">
        
        <div class="method-heading">
          <span class="method-name">fixmate</span><span
            class="method-args">(opts={})</span>
          
        </div>
        

        <div class="method-description">
          
          
          
          

          
        </div>

        

        
        <div class="aliases">
          Alias for: <a href="Sam.html#method-i-fix_mates">fix_mates</a>
        </div>
        
      </div><!-- fixmate-method -->

    
      <div id="method-i-flag_stats" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">flag_stats</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>generate simple stats with regard to the number and pairing of reads mapped
to a reference</p>
          
          

          
          <div class="method-source-code" id="flag_stats-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 262</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">flag_stats</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&quot;flagstat&quot;</span>, <span class="ruby-identifier">opts</span>, [])
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">strings</span> = []
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>,<span class="ruby-constant">String</span>) {<span class="ruby-operator">|</span><span class="ruby-identifier">line</span><span class="ruby-operator">|</span> <span class="ruby-identifier">strings</span> <span class="ruby-operator">&lt;&lt;</span> <span class="ruby-identifier">line</span>.<span class="ruby-identifier">chomp</span>}
  <span class="ruby-identifier">strings</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- flag_stats-source -->
          
        </div>

        
        <div class="aliases">
          Also aliased as: <a href="Sam.html#method-i-flagstat">flagstat</a>
        </div>
        

        
      </div><!-- flag_stats-method -->

    
      <div id="method-i-flagstat" class="method-detail method-alias">
        
        <div class="method-heading">
          <span class="method-name">flagstat</span><span
            class="method-args">(opts={})</span>
          
        </div>
        

        <div class="method-description">
          
          
          
          

          
        </div>

        

        
        <div class="aliases">
          Alias for: <a href="Sam.html#method-i-flag_stats">flag_stats</a>
        </div>
        
      </div><!-- flagstat-method -->

    
      <div id="method-i-idxstats" class="method-detail method-alias">
        
        <div class="method-heading">
          <span class="method-name">idxstats</span><span
            class="method-args">()</span>
          
        </div>
        

        <div class="method-description">
          
          
          
          

          
        </div>

        

        
        <div class="aliases">
          Alias for: <a href="Sam.html#method-i-index_stats">index_stats</a>
        </div>
        
      </div><!-- idxstats-method -->

    
      <div id="method-i-index" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">index</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Index sorted alignment for fast random access. Index file
&lt;aln.bam&gt;.bai will be created of no out_index is provided.</p>
<ul><li>
<p>out_index - [STRING] name of index</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="index-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 237</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">index</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} index #{@bam} #{opts[:out_index]}&quot;</span>
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- index-source -->
          
        </div>

        

        
      </div><!-- index-method -->

    
      <div id="method-i-index_stats" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">index_stats</span><span
            class="method-args">()</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Retrieve and print stats in the index file. The output is TAB delimited
with each line consisting of reference sequence name, sequence length,
number of mapped reads and number unmapped reads.</p>
          
          

          
          <div class="method-source-code" id="index_stats-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 274</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">index_stats</span>
  <span class="ruby-identifier">stats</span> = {}   
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-node">&quot;idxstats #{@bam}&quot;</span>, {}, [])
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-constant">String</span>, <span class="ruby-value">:text</span>, <span class="ruby-keyword">true</span>, <span class="ruby-node">&quot;#&quot;</span>) <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">line</span><span class="ruby-operator">|</span>
    <span class="ruby-identifier">info</span> = <span class="ruby-identifier">line</span>.<span class="ruby-identifier">chomp</span>.<span class="ruby-identifier">split</span>(<span class="ruby-regexp">/\t/</span>)
    <span class="ruby-identifier">stats</span>[ <span class="ruby-identifier">info</span>[<span class="ruby-value">0</span>] ] = {<span class="ruby-value">:length</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">info</span>[<span class="ruby-value">1</span>].<span class="ruby-identifier">to_i</span>, <span class="ruby-value">:mapped_reads</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">info</span>[<span class="ruby-value">2</span>].<span class="ruby-identifier">to_i</span>, <span class="ruby-value">:unmapped_reads</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-identifier">info</span>[<span class="ruby-value">3</span>].<span class="ruby-identifier">to_i</span> }
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">stats</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- index_stats-source -->
          
        </div>

        
        <div class="aliases">
          Also aliased as: <a href="Sam.html#method-i-idxstats">idxstats</a>
        </div>
        

        
      </div><!-- index_stats-method -->

    
      <div id="method-i-merge" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">merge</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Merge multiple sorted alignments</p>
<ul><li>
<p>n - sort by read names</p>
</li><li>
<p>r - attach RG tag (inferred from file names)</p>
</li><li>
<p>u - uncompressed BAM output</p>
</li><li>
<p>f - overwrite the output BAM if exist</p>
</li><li>
<p>one - compress level 1</p>
</li><li>
<p>l  - [INT] compression level, from 0 to 9 [-1]</p>
</li><li>
<p>at - [INT] number of BAM compression threads [0]</p>
</li><li>
<p>R - [STRING] merge file in the specified region STR [all]</p>
</li><li>
<p>h - [FILE] copy the header in FILE to &lt;out.bam&gt; [in1.bam]</p>
</li><li>
<p>out - [FILE] out file name</p>
</li><li>
<p>bams - [FILES] or <a href="Sam.html">Bio::DB::Sam</a> list of input bams,
or <a href="Sam.html">Bio::DB::Sam</a> objects</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="merge-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 299</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">merge</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:one</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;1&#39;</span>] = <span class="ruby-keyword">nil</span>
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:one</span>)
  <span class="ruby-keyword">end</span>

  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:at</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;@&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:at</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:at</span>)
  <span class="ruby-keyword">end</span>

  <span class="ruby-identifier">out</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:out</span>]
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:out</span>)

  <span class="ruby-identifier">bam_list</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:bams</span>].<span class="ruby-identifier">collect</span> <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">b</span><span class="ruby-operator">|</span>
    <span class="ruby-identifier">b</span>.<span class="ruby-identifier">bam</span> <span class="ruby-keyword">rescue</span> <span class="ruby-identifier">b</span>
  <span class="ruby-keyword">end</span>.<span class="ruby-identifier">join</span>(<span class="ruby-string">&#39; &#39;</span>)

  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:bams</span>)
  <span class="ruby-identifier">options</span> = <span class="ruby-identifier">commandify</span>(<span class="ruby-identifier">opts</span>, [<span class="ruby-value">:n</span>, <span class="ruby-value">:r</span>, <span class="ruby-value">:u</span>, <span class="ruby-value">:f</span>, <span class="ruby-string">&#39;1&#39;</span>] )
  <span class="ruby-comment">#command = &quot;#{form_opt_string(@samtools, &quot;merge&quot;, opts, [:n, :r, :u, :f, &#39;1&#39;] )} #{out} #{bam_list}&quot;</span>
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} merge #{options} #{out} #{bam_list}&quot;</span>

  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">puts</span> <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)

<span class="ruby-keyword">end</span></pre>
          </div><!-- merge-source -->
          
        </div>

        

        
      </div><!-- merge-method -->

    
      <div id="method-i-mpileup" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">mpileup</span><span
            class="method-args">(opts={}, &block)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>returns a <a href="Pileup.html">Bio::DB::Pileup</a> or Bio::DB::VCF object</p>
<ul><li>
<p>region - Only generate pileup in region [chrom:start-stop]</p>
</li><li>
<p>illumina_quals - Assume the quality is in the Illumina 1.3+ encoding</p>
</li><li>
<p>count_anomalous - Do not skip anomalous read pairs in variant calling</p>
</li><li>
<p>no_baq - Disable probabilistic realignment for the computation of base
alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base
being misaligned. Applying this option greatly helps to reduce false SNPs
caused by misalignments.</p>
</li><li>
<p>adjust_mapq - [INT] Coefficient for downgrading mapping quality for reads
containing excessive mismatches. Given a read with a phred-scaled
probability q of being generated from the mapped position, the new mapping
quality is about sqrt((INT-q)/INT)*INT. A zero value disables this
functionality; if enabled, the recommended value for BWA is 50. [0]</p>
</li><li>
<p>max_per_bam_depth - [INT] At a position, read maximally INT reads per input
BAM. [250]</p>
</li><li>
<p>extended_baq - Extended BAQ computation. This option helps sensitivity
especially for MNPs, but may hurt specificity a little bit.</p>
</li><li>
<p>exclude_reads_file - [FILE] exclude read groups listed in FILE [null]</p>
</li><li>
<p>list_of_positions - [FILE] BED or position list file containing a list of
regions or sites where pileup or BCF should be generated [null]</p>
</li><li>
<p>mapping_quality_cap - [INT] cap mapping quality at INT [60]</p>
</li><li>
<p>ignore_rg - ignore read group tags</p>
</li><li>
<p>min_mapping_quality - [INT] skip alignments with mapQ smaller than INT [0]</p>
</li><li>
<p>min_base_quality - [INT] skip bases with baseQ/BAQ smaller than INT [13]</p>
</li><li>
<p>###following options are for the -g -u option</p>
</li><li>
<p>genotype_calling - generate BCF output (genotype likelihoods)</p>
</li><li>
<p>uncompressed_bcf - generate uncompress BCF output</p>
</li><li>
<p>extension_sequencing_probability - [INT] Phred-scaled gap extension seq
error probability [20]</p>
</li><li>
<p>homopolymer_error_coefficient - [INT] coefficient for homopolymer errors
[100]</p>
</li><li>
<p>no_indels - do not perform indel calling</p>
</li><li>
<p>skip_indel_over_average_depth - [INT] max per-sample depth for INDEL
calling [250]</p>
</li><li>
<p>gap_open_sequencing_error_probability - [INT] Phred-scaled gap open
sequencing error probability [40]</p>
</li><li>
<p>platforms - [STRING] comma separated list of platforms for indels [all]</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="mpileup-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 144</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">mpileup</span>(<span class="ruby-identifier">opts</span>={}, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
  <span class="ruby-comment">#long option form to short samtools form..</span>
  <span class="ruby-identifier">long_opts</span> = {
  <span class="ruby-value">:region</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:r</span>,
  <span class="ruby-value">:illumina_quals</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:six</span>,
  <span class="ruby-value">:count_anomalous</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:A</span>,
  <span class="ruby-value">:no_baq</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:B</span>,
  <span class="ruby-value">:adjust_mapq</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:C</span>,
  <span class="ruby-value">:max_per_bam_depth</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:d</span>,
  <span class="ruby-value">:extended_baq</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:E</span>,
  <span class="ruby-value">:exclude_reads_file</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:G</span>,
  <span class="ruby-value">:list_of_positions</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:l</span>,
  <span class="ruby-value">:mapping_quality_cap</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:M</span>,
  <span class="ruby-value">:ignore_rg</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:R</span>,
  <span class="ruby-value">:min_mapping_quality</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:q</span>,
  <span class="ruby-value">:min_base_quality</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:Q</span>,
  <span class="ruby-comment">###following options are for the -g -u option</span>
  <span class="ruby-value">:genotype_calling</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:g</span>,
  <span class="ruby-value">:uncompressed_bcf</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:u</span>,
  <span class="ruby-value">:extension_sequencing_probability</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:e</span>,
  <span class="ruby-value">:homopolymer_error_coefficient</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:h</span>,
  <span class="ruby-value">:no_indels</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:I</span>,
  <span class="ruby-value">:skip_indel_over_average_depth</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:L</span>,
  <span class="ruby-value">:gap_open_sequencing_error_probability</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:o</span>,
  <span class="ruby-value">:platforms</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">:P</span> 
  }

  <span class="ruby-comment">##convert any long_opts to short opts </span>
  <span class="ruby-identifier">temp_opts</span> = <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">dup</span>
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">each_pair</span> <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">k</span>,<span class="ruby-identifier">v</span><span class="ruby-operator">|</span>
    <span class="ruby-keyword">if</span> <span class="ruby-identifier">long_opts</span>[<span class="ruby-identifier">k</span>]
      <span class="ruby-identifier">temp_opts</span>[<span class="ruby-identifier">long_opts</span>[<span class="ruby-identifier">k</span>]] = <span class="ruby-identifier">v</span> 
      <span class="ruby-identifier">temp_opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-identifier">k</span>)
    <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">opts</span> = <span class="ruby-identifier">temp_opts</span>
  <span class="ruby-identifier">opts</span>[<span class="ruby-value">:u</span>] = <span class="ruby-keyword">true</span> <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:g</span>] <span class="ruby-comment">#so that we always get uncompressed output</span>
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:g</span>)
  
  <span class="ruby-identifier">opts</span>[<span class="ruby-value">:f</span>] = <span class="ruby-ivar">@fasta</span>
  
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:six</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&quot;6&quot;</span>] = <span class="ruby-keyword">nil</span>
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:six</span>)
  <span class="ruby-keyword">end</span>
  
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&quot;mpileup&quot;</span>, <span class="ruby-identifier">opts</span>, [<span class="ruby-value">:R</span>, <span class="ruby-value">:B</span>, <span class="ruby-value">:E</span>, <span class="ruby-string">&quot;6&quot;</span>, <span class="ruby-value">:A</span>, <span class="ruby-value">:g</span>, <span class="ruby-value">:u</span>, <span class="ruby-value">:I</span>] )

  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:u</span>]
    <span class="ruby-identifier">command</span> = <span class="ruby-identifier">command</span> <span class="ruby-operator">+</span> <span class="ruby-node">&quot; | #{@bcftools} view -cg -&quot;</span>
  <span class="ruby-keyword">end</span>
  
  <span class="ruby-identifier">klass</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:u</span>] <span class="ruby-operator">?</span> <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">DB</span><span class="ruby-operator">::</span><span class="ruby-constant">Vcf</span> <span class="ruby-operator">:</span> <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">DB</span><span class="ruby-operator">::</span><span class="ruby-constant">Pileup</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-identifier">klass</span>, <span class="ruby-value">:text</span>, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)

<span class="ruby-keyword">end</span></pre>
          </div><!-- mpileup-source -->
          
        </div>

        

        
      </div><!-- mpileup-method -->

    
      <div id="method-i-open" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">open</span><span
            class="method-args">()</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>backward compatibility method, returns true if file exists otherwise,
complains and quits.</p>
          
          

          
          <div class="method-source-code" id="open-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 28</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">open</span>
  <span class="ruby-identifier">files_ok?</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- open-source -->
          
        </div>

        

        
      </div><!-- open-method -->

    
      <div id="method-i-phase" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">phase</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>A - Drop reads with ambiguous phase. b - [STR] Prefix of BAM output. When
this option is in use, phase-0 reads will be saved in file STR.0.bam and
phase-1 reads in STR.1.bam. Phase unknown reads will be randomly allocated
to one of the two files. Chimeric reads with switch errors will be saved in
STR.chimeric.bam. [null] F - Do not attempt to fix chimeric reads. k -
[INT] Maximum length for local phasing. [13] q - [INT] Minimum Phred-scaled
LOD to call a heterozygote. [40] Q - [INT] Minimum base quality to be used
in het calling. [13]</p>
          
          

          
          <div class="method-source-code" id="phase-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 467</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">phase</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{form_opt_string(@samtools, &quot;phase&quot;, opts, [:A, :F] )}&quot;</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- phase-source -->
          
        </div>

        

        
      </div><!-- phase-method -->

    
      <div id="method-i-reheader" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">reheader</span><span
            class="method-args">(header_sam, opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>out - [FILE] output bam file</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="reheader-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 414</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">reheader</span>(<span class="ruby-identifier">header_sam</span>, <span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">has_key?</span>(<span class="ruby-value">:out</span>)
    <span class="ruby-identifier">out</span>=<span class="ruby-identifier">opts</span>[<span class="ruby-value">:out</span>]
    <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} reheader #{header_sam} #{@bam} &gt; #{out}&quot;</span>
  <span class="ruby-keyword">else</span>
    <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{@samtools} reheader #{header_sam} #{@bam}&quot;</span>
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">puts</span> <span class="ruby-identifier">command</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- reheader-source -->
          
        </div>

        

        
      </div><!-- reheader-method -->

    
      <div id="method-i-remove_duplicates" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">remove_duplicates</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>s - rmdup for SE reads</p>
</li><li>
<p>S - treat PE reads as SE in rmdup (force -s)</p>
</li><li>
<p>out - [FILE] output bam</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="remove_duplicates-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 358</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">remove_duplicates</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-identifier">out</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:out</span>]
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:out</span>)
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{form_opt_string(@samtools, &quot;rmdup&quot;, opts, [:s, :S])} #{out} #{@bam}&quot;</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- remove_duplicates-source -->
          
        </div>

        
        <div class="aliases">
          Also aliased as: <a href="Sam.html#method-i-rmdup">rmdup</a>
        </div>
        

        
      </div><!-- remove_duplicates-method -->

    
      <div id="method-i-rmdup" class="method-detail method-alias">
        
        <div class="method-heading">
          <span class="method-name">rmdup</span><span
            class="method-args">(opts={})</span>
          
        </div>
        

        <div class="method-description">
          
          
          
          

          
        </div>

        

        
        <div class="aliases">
          Alias for: <a href="Sam.html#method-i-remove_duplicates">remove_duplicates</a>
        </div>
        
      </div><!-- rmdup-method -->

    
      <div id="method-i-sort" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">sort</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>n - sort by read name</p>
</li><li>
<p>f - use &lt;out.prefix&gt; as full file name instead of prefix</p>
</li><li>
<p>o - final output to stdout returns bio::db::alignment</p>
</li><li>
<p>l - [INT]  compression level, from 0 to 9 [-1]</p>
</li><li>
<p>at - [INT] number of sorting and compression threads [1]</p>
</li><li>
<p>m - [INT] max memory per thread; suffix K/M/G recognized [768M]</p>
</li><li>
<p>prefix - [STRING] prefix for output bamfile</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="sort-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 375</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">sort</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-operator">!</span><span class="ruby-identifier">opts</span>.<span class="ruby-identifier">has_key?</span>(<span class="ruby-value">:prefix</span>)
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">merge!</span>({<span class="ruby-value">:prefix</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-string">&quot;sorted&quot;</span>})
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">prefix</span> = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:prefix</span>]
  <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:prefix</span>)
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&quot;sort&quot;</span>, <span class="ruby-identifier">opts</span>, [<span class="ruby-value">:n</span>, <span class="ruby-value">:f</span>, <span class="ruby-value">:o</span>])
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">command</span> <span class="ruby-operator">+</span> <span class="ruby-string">&quot; &quot;</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">prefix</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-comment">#puts command</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:o</span>]
    <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">DB</span><span class="ruby-operator">::</span><span class="ruby-constant">Alignment</span>)
  <span class="ruby-keyword">else</span>
    <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
  <span class="ruby-keyword">end</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- sort-source -->
          
        </div>

        

        
      </div><!-- sort-method -->

    
      <div id="method-i-targetcut" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">targetcut</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>Identifies target regions by examining the continuity of read depth,
computes haploid consensus sequences of targets and outputs a <a
href="SAM.html">SAM</a> with each sequence corresponding to a target. When
option -f is in use, BAQ will be applied.</p>
<ul><li>
<p>Q - [INT] Minimum base quality for a base to be considered [13]</p>
</li><li>
<p>i - in penalty</p>
</li><li>
<p>0 - em0</p>
</li><li>
<p>1 - em1</p>
</li><li>
<p>2 - em2</p>
</li><li>
<p>f - reference</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="targetcut-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 450</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">targetcut</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:f</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;f&#39;</span>] = <span class="ruby-ivar">@fasta</span>
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:s</span>)
  <span class="ruby-keyword">end</span>
  
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{form_opt_string(@samtools, &quot;targetcut&quot;, opts, [] )}&quot;</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- targetcut-source -->
          
        </div>

        

        
      </div><!-- targetcut-method -->

    
      <div id="method-i-tview" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">tview</span><span
            class="method-args">(opts={})</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <ul><li>
<p>d - display, output as (H)tml or (C)urses or (T)ext</p>
</li><li>
<p>p - [chr:pos] go directly to this position</p>
</li><li>
<p>s - [STR] display only reads from this sample or group</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="tview-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 395</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">tview</span>(<span class="ruby-identifier">opts</span>={})
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:d</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;d&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:d</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:d</span>)
  <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:p</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;p&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:p</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:p</span>)
  <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:s</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;s&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:s</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:s</span>)
  <span class="ruby-keyword">end</span>
  <span class="ruby-identifier">command</span> = <span class="ruby-node">&quot;#{form_opt_string(@samtools, &quot;tview&quot;, opts)}&quot;</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">system</span>(<span class="ruby-identifier">command</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- tview-source -->
          
        </div>

        

        
      </div><!-- tview-method -->

    
      <div id="method-i-view" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">view</span><span
            class="method-args">(opts={},&block)</span>
          
          <span class="method-click-advice">click to toggle source</span>
          
        </div>
        

        <div class="method-description">
          
          <p>runs the samtools view command @param opts [Hash] options for view as
follows</p>

<h2 id="method-i-view-label-args">args<span><a href="#method-i-view-label-args">&para;</a> <a href="#documentation">&uarr;</a></span></h2>
<ul><li>
<p>b - output BAM</p>
</li><li>
<p>h - print header for the <a href="SAM.html">SAM</a> output</p>
</li><li>
<p>H - print header only (no alignments)</p>
</li><li>
<p>S - input is <a href="SAM.html">SAM</a></p>
</li><li>
<p>u - uncompressed BAM output (force -b)</p>
</li><li>
<p>one - fast compression (force -b)</p>
</li><li>
<p>x - output FLAG in HEX (samtools-C specific)</p>
</li><li>
<p>X - output FLAG in string (samtools-C specific)</p>
</li><li>
<p>c - print only the count of matching records</p>
</li><li>
<p>B - collapse the backward CIGAR operation</p>
</li><li>
<p>at - INT number of BAM compression threads [0]</p>
</li><li>
<p>L - FILE output alignments overlapping the input BED FILE [null]</p>
</li><li>
<p>t - FILE list of reference names and lengths (force -S) [null]</p>
</li><li>
<p>T - FILE reference sequence file (force -S) [null]</p>
</li><li>
<p>o - FILE output file name [stdout]</p>
</li><li>
<p>R - FILE list of read groups to be outputted [null]</p>
</li><li>
<p>f - INT required flag  0 for unset [0]</p>
</li><li>
<p>F - INT filtering flag  0 for unset [0]</p>
</li><li>
<p>q - INT minimum mapping quality [0]</p>
</li><li>
<p>l - STR only output reads in library STR [null]</p>
</li><li>
<p>r - STR only output reads in read group STR [null]</p>
</li><li>
<p>s - FLOAT fraction of templates to subsample; integer part as seed [-1]</p>
</li><li>
<p>chr - name of reference sequence to get alignments from</p>
</li><li>
<p>start - start position on reference sequence</p>
</li><li>
<p>stop - end postion on reference sequence</p>
</li></ul>
          
          

          
          <div class="method-source-code" id="view-source">
            <pre><span class="ruby-comment"># File lib/bio/db/sam.rb, line 60</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">view</span>(<span class="ruby-identifier">opts</span>={},<span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
  <span class="ruby-identifier">region</span> = <span class="ruby-constant">String</span>.<span class="ruby-identifier">new</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:chr</span>] <span class="ruby-keyword">and</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:start</span>] <span class="ruby-keyword">and</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:stop</span>]
    <span class="ruby-identifier">region</span> = <span class="ruby-node">&quot;#{opts[:chr]}:#{opts[:start]}-#{opts[:stop]}&quot;</span>
    [<span class="ruby-value">:chr</span>, <span class="ruby-value">:start</span>, <span class="ruby-value">:stop</span>].<span class="ruby-identifier">each</span> {<span class="ruby-operator">|</span><span class="ruby-identifier">o</span><span class="ruby-operator">|</span> <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-identifier">o</span>)}
  <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:at</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;@&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:at</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:at</span>)
  <span class="ruby-keyword">end</span>
  
  <span class="ruby-keyword">if</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:one</span>]
    <span class="ruby-identifier">opts</span>[<span class="ruby-string">&#39;1&#39;</span>] = <span class="ruby-identifier">opts</span>[<span class="ruby-value">:one</span>]
    <span class="ruby-identifier">opts</span>.<span class="ruby-identifier">delete</span>(<span class="ruby-value">:one</span>)
  <span class="ruby-keyword">end</span>
          
  <span class="ruby-identifier">command</span> = <span class="ruby-identifier">form_opt_string</span>(<span class="ruby-ivar">@samtools</span>, <span class="ruby-string">&#39;view&#39;</span>, <span class="ruby-identifier">opts</span>, [<span class="ruby-value">:b</span>, <span class="ruby-value">:h</span>, <span class="ruby-value">:H</span>, <span class="ruby-value">:S</span>, <span class="ruby-value">:u</span>, <span class="ruby-string">&#39;1&#39;</span>, <span class="ruby-value">:x</span>, <span class="ruby-value">:X</span>, <span class="ruby-value">:c</span>, <span class="ruby-value">:B</span>]) <span class="ruby-operator">+</span> <span class="ruby-string">&quot; &quot;</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">region</span>
  <span class="ruby-ivar">@last_command</span> = <span class="ruby-identifier">command</span>
  <span class="ruby-identifier">type</span> = (<span class="ruby-identifier">opts</span>[<span class="ruby-value">:u</span>] <span class="ruby-keyword">or</span> <span class="ruby-identifier">opts</span>[<span class="ruby-value">:b</span>]) <span class="ruby-operator">?</span> <span class="ruby-value">:binary</span> <span class="ruby-operator">:</span> <span class="ruby-value">:text</span>
  <span class="ruby-identifier">klass</span> = (<span class="ruby-identifier">type</span> <span class="ruby-operator">==</span> <span class="ruby-value">:binary</span>) <span class="ruby-operator">?</span> <span class="ruby-constant">String</span> <span class="ruby-operator">:</span> <span class="ruby-constant">Bio</span><span class="ruby-operator">::</span><span class="ruby-constant">DB</span><span class="ruby-operator">::</span><span class="ruby-constant">Alignment</span>
  <span class="ruby-identifier">yield_from_pipe</span>(<span class="ruby-identifier">command</span>, <span class="ruby-identifier">klass</span>, <span class="ruby-identifier">type</span>, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
<span class="ruby-keyword">end</span></pre>
          </div><!-- view-source -->
          
        </div>

        

        
      </div><!-- view-method -->

    
    </section><!-- public-instance-method-details -->
  
  </section><!-- 5Buntitled-5D -->

</div><!-- documentation -->


<footer id="validator-badges">
  <p><a href="http://validator.w3.org/check/referer">[Validate]</a>
  <p>Generated by <a href="https://github.com/rdoc/rdoc">RDoc</a> 4.0.1.
  <p>Generated with the <a href="http://deveiate.org/projects/Darkfish-Rdoc/">Darkfish Rdoc Generator</a> 3.
</footer>

